webserver/static/js/face.ts

```diff
@@ -0,0 +1,698 @@
+    );
+  }
+
+  increasingState(): boolean {
+    return (
+      (this.state() == Gradient.Increasing &&
+        this.previousState() >= Gradient.Neutral) ||
+      (this.state() == Gradient.Neutral &&
+        this.previousState() == Gradient.Increasing)
+    );
+  }
+  previousState(): Gradient {
+    if (this.states.values.length > 1) {
+      return this.states.values[this.states.values.length - 2];
+    }
+    return 0;
+  }
+
+  state(): Gradient {
+    return this.states.values[this.states.values.length - 1];
+  }
+
+  // allow for 1 of the last 2 values to not be increasing i.e. noise
+  increasing(): boolean {
+    if (this.deltas.values.length < 3) {
+      return false;
+    }
+    return (
+      this.deltas.values[2] > 0 &&
+      (this.deltas.values[1] > 0 || this.deltas.values[0] > 0)
+    );
+  }
+
+  // allow for 1 of the last 2 values to not be decreasing i.e. noise
+  decreasing(): boolean {
+    if (this.deltas.values.length < 3) {
+      return false;
+    }
+    return (
+      this.deltas.values[2] < 0 &&
+      (this.deltas.values[1] < 0 || this.deltas.values[0] < 0)
+    );
+  }
+}
+
+export class Hotspot {
+  sensorValue: number;
+  sensorX: number;
+  sensorY: number;
+  sensorCorrection: number;
+  constructor() {
+    this.sensorCorrection = 0;
+    this.sensorValue = 0;
+    this.sensorX = 0;
+    this.sensorY = 0;
+  }
+}
+export class Face {
+  numFrames: number;
+  framesMissing: number;
+  id: number;
+  hotspot: Hotspot;
+  roi: ROIFeature | null;
+  forehead: ROIFeature | null;
+  haarAge: number;
+  haarLastSeen: number;
+  xFeatures: ROIFeature[];
+  yFeatures: ROIFeature[];
+  widthWindow: Window;
+  heightWindow: Window;
+  faceWidth: number;
+  faceHeight: number;
+  constructor(public haarFace: ROIFeature, public frameTime: number) {
+    this.roi = null;
+    this.forehead = null;
+    this.id = 0;
+    this.framesMissing = 0;
+    this.haarAge = 1;
+    this.haarLastSeen = 0;
+    this.hotspot = new Hotspot();
+    this.numFrames = 0;
+    this.widthWindow = new Window(3);
+    this.heightWindow = new Window(3);
+    this.faceWidth = 0;
+    this.faceHeight = 0;
+    //debugging
+    this.xFeatures = [];
+    this.yFeatures = [];
+    this.assignID();
+  }
+
+  updateHaar(haar: ROIFeature) {
+    this.haarFace = haar;
+    this.haarAge++;
+    this.haarLastSeen = this.numFrames + 1;
+  }
+
+  haarActive(): boolean {
+    return this.haarLastSeen == this.numFrames;
+  }
+
+  tracked(): boolean {
+    return this.roi != null && this.forehead != null;
+  }
+
+  assignID() {
+    this.id = FaceID;
+    FaceID++;
+  }
+
+  height(): number {
+    if (this.roi) {
+      return this.roi.height();
+    }
+    return 0;
+  }
+
+  width(): number {
+    if (this.roi) {
+      return this.roi.width();
+    }
+    return 0;
+  }
+
+  active(): boolean {
+    return this.framesMissing < maxFrameSkip;
+  }
+
+  update(forehead: ROIFeature | null, roi: ROIFeature | null) {
```

Comment:
Typing of this seems problematic, if these args can indeed be null, then how can you assume they're not null in the body of the function?

Reply:
thats what this.tracked() does